experimental/ir/fdp.go

```diff
@@ -0,0 +1,443 @@
+	var c codegen
+	for _, opt := range options {
+		if opt != nil {
+			opt(&c)
+		}
+	}
+
+	fdp := new(descriptorpb.FileDescriptorProto)
+	c.file(file, fdp)
+	return proto.Marshal(fdp)
+}
+
+// CodegenOption is an option to pass to [DescriptorSet].
+type CodegenOption func(*codegen)
+
+// IncludeWKTs is an option for including FDPs for the Well-Known Types in
+// codegen output.
+func IncludeWKTs() CodegenOption {
+	return func(c *codegen) { c.includeWKTs = true }
+}
+
+// IncludeSourceCodeInfo is an option that requires generating source code info
+// for the descriptors.
+func IncludeSourceCodeInfo() CodegenOption {
+	return func(c *codegen) { c.includeSourceCodeInfo = true }
+}
+
+type codegen struct {
+	includeWKTs           bool
+	includeSourceCodeInfo bool
+
+	currentFile File
+}
+
+func (c *codegen) files(files []File, fds *descriptorpb.FileDescriptorSet) {
+	if c.includeSourceCodeInfo {
+		panic("protocompile/ir: NYI: IncludeSourceCodeInfo()")
+	}
+
+	// Build up all of the imported files. We can't just pull out the transitive
+	// imports for each file because we want the result to be sorted
+	// topologically.
+	for file := range TopoSort(files...) {
+		if !c.includeWKTs && wellknownimports.StandardImport(file.Path()) != "" {
+			continue
+		}
+
+		fdp := new(descriptorpb.FileDescriptorProto)
+		fds.File = append(fds.File, fdp)
+		c.file(file, fdp)
+	}
+}
+
+func (c *codegen) file(file File, fdp *descriptorpb.FileDescriptorProto) {
+	c.currentFile = file
+
+	fdp.Name = addr(file.Path())
+	fdp.Package = addr(file.Context().Package())
+
+	switch file.Syntax() {
+	case syntax.Proto2, syntax.Proto3:
+		fdp.Syntax = addr(file.Syntax().String())
+	case syntax.Edition2023:
+		fdp.Syntax = addr("editions")
+		fdp.Edition = descriptorpb.Edition_EDITION_2023.Enum()
+	}
+
+	for imp := range seq.Values(file.Imports()) {
+		fdp.Dependency = append(fdp.Dependency, imp.Path())
+		if imp.Public {
+			fdp.PublicDependency = append(fdp.PublicDependency, int32(len(fdp.Dependency)-1))
+		}
+		if imp.Weak {
+			fdp.WeakDependency = append(fdp.WeakDependency, int32(len(fdp.Dependency)-1))
+		}
+	}
+
+	for ty := range seq.Values(file.Types()) {
+		if ty.IsEnum() {
+			edp := new(descriptorpb.EnumDescriptorProto)
+			fdp.EnumType = append(fdp.EnumType, edp)
+			c.enum(ty, edp)
+			continue
+		}
+
+		mdp := new(descriptorpb.DescriptorProto)
+		fdp.MessageType = append(fdp.MessageType, mdp)
+		c.message(ty, mdp)
+	}
+
+	// TODO: Services.
+
+	for extn := range seq.Values(file.Extensions()) {
+		fd := new(descriptorpb.FieldDescriptorProto)
+		fdp.Extension = append(fdp.Extension, fd)
+		c.field(extn, fd)
+	}
+
+	for option := range seq.Values(file.Options()) {
+		if fdp.Options == nil {
+			fdp.Options = new(descriptorpb.FileOptions)
+		}
+
+		c.option(option, fdp.Options)
+	}
+}
+
+func (c *codegen) message(ty Type, mdp *descriptorpb.DescriptorProto) {
+	mdp.Name = addr(ty.Name()) // Has a leading dot.
```

Comment:
I don't understand the comment. `DescriptorProto.Name` should not have a leading dot and, in fact, should not be qualified at all. Qualifying names is only done for references -- like field type names, message type names for methods, etc.

Reply:
Facepalm